server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -63,6 +73,43 @@ protected void unpromotableShardOperation(
         UnpromotableShardRefreshRequest request,
         ActionListener<ActionResponse.Empty> responseListener
     ) {
+        if (clusterService.state().version() >= request.getClusterStateVersion()) {
+            doUnpromotableShardOperation(request, responseListener);
+        } else {
+            final Thread initialThread = Thread.currentThread();
+            ClusterStateObserver.waitForState(
+                clusterService,
+                clusterService.threadPool().getThreadContext(),
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        if (initialThread == Thread.currentThread()) {
```

Comment:
nit Would it be more efficient to check whether it's the same executor as before? I.e., if it's another thread, but in the same executor pool, we call `doUnpromotableShardOperation` directly.